README.md
# Django Blacklist

Blacklist users and hosts in Django. Automatically blacklist rate-limited clients.


## Overview

`Django Blacklist` allows you to block specific users and IP addresses/networks from accessing your application.
Clients can be blocked manually from the admin interface, or automatically after exceeding a request rate limit.
Each blacklist rule is applied for a specific duration.
The blacklist is very scalable and is applied without noticeable overhead for large numbers of rules.


## Installation

To install the package, run:
```
$ pip install django-blacklist
```

Add the `blacklist` application to `INSTALLED_APPS`:
```
INSTALLED_APPS = [
    ...
    'blacklist'
]
```

Add the `BlacklistMiddleware` middleware after `AuthenticationMiddleware`:
```
MIDDLEWARE = [
    ...
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'blacklist.middleware.BlacklistMiddleware',
    ...
]
```

Apply the blacklist database migrations:
```
$ python manage.py migrate blacklist
```


## Usage

You can manage the blacklist rules from the admin.
Changes take effect after a configurable time, or when the server is restarted.
A rule can target a user or an IP address.
You can also target IP networks (ranges) by specifying the optional `prefixlen` field (number of network prefix bits).
Each rule has a specific duration. After that time interval passes, the rule expires.
When a request is blocked due to a matching rule:
* Status 400 (bad request) is returned.
* An error template is rendered.
  You can specify a custom one (see `Settings` below), or use the one for status 400.
* A message is logged
  (warning from logger `blacklist.middleware` for custom templates, or error from logger `django.security` otherwise).

### Removing Expired Rules

Expired rules are not automatically removed from the database.
They can be cleaned up with the included management command `trim_blacklist`:
```
$ python manage.py trim_blacklist [-c <created_days>] [-e <expired_days>]
```
The options `-c` and `-e` specify the minimum ages of creation and expiry, respectively.


## Automatic Blacklisting

Clients can be blacklisted automatically, after exceeding a specified request rate limit.
This feature requires [django-ratelimit](https://github.com/jsocol/django-ratelimit).

First, rate-limit a view by applying the `@ratelimit` decorator. Make sure to set `block=False`.
Then, blacklist rate-limited clients by adding the `@blacklist_ratelimited` decorator. Specify the blacklist duration.
For example:
```
from datetime import timedelta
from django_ratelimit.decorators import ratelimit
from blacklist.ratelimit import blacklist_ratelimited

@ratelimit(key='user_or_ip', rate='50/m', block=False)
@blacklist_ratelimited(timedelta(minutes=30))
def index(request):
    ...
```

Automatic rules take effect immediately.
If the request comes from an authenticated user, the rule will target that user.
Otherwise, it will target their IP address.

`@blacklist_ratelimited` accepts two arguments: `(duration, block=True)`.
* `duration` can be a `timedelta` object, or a tuple of two separate durations
(for user-based and IP-based rules).
* `block` specifies if the request should be blocked immediately, or passed to the view.

Automatic rules will have a comment that contains the ID of the request, which triggered the creation of the rule,
and the "request line".
The request ID is added only if available. Django does not generate request IDs.
For that purpose, you can install [django-log-request-id](https://github.com/dabapps/django-log-request-id).


## Proxies and Client Addresses

By default, the client IP address is taken from the `REMOTE_ADDR` value of `request.META`.
If your application server is behind one or more reverse proxies,
this will usually be the address of the nearest proxy, and not the actual client address.
To properly blacklist clients by IP address,
you can configure `Django Blacklist` to use addresses from another source (see `Settings` below).

To actually obtain the proxied client addresses,
you can use [django-ipware](https://github.com/un33k/django-ipware).
In this case, you can configure `Django Blacklist` to obtain client addresses from your function,
which in turn calls `django-ipware` for the actual logic.

Alternatively, you can set `REMOTE_ADDR` from the `X-Forwarded-For` header in middleware,
installed before `Django Blacklist`.
However, keep in mind that this header can be forged to bypass the rate limits.
To counter that, you can use the last address in that header (which should be set by your trusted reverse proxy).
If you are behind two proxies, use the second to last address, and so on.


## Settings

* `BLACKLIST_ENABLE` - whether blacklisted clients should be blocked,
  and rate-limited clients should be blacklisted; default: `True`
* `BLACKLIST_RELOAD_PERIOD` - how often to reload the blacklist, in seconds; default: `60`
* `BLACKLIST_RATELIMITED_ENABLE` - whether rate-limited clients should be automatically blacklisted;
  requires `BLACKLIST_ENABLE`; default: `True`
* `BLACKLIST_TEMPLATE` - name of a custom error template to render to blocked clients;
  its context will contain `request` and `exception`;
  set to `None` to use the template for status 400; default: `None`
* `BLACKLIST_LOGGING_ENABLE` - whether blocked requests should be logged
  (honored only if a custom error template is configured); default: `True`
* `BLACKLIST_ADDRESS_SOURCE` - the source of client addresses; can be a key in `request.META`,
  a callable that receives the request object, or the dotted string path to such a callable;
  default: `'REMOTE_ADDR'`


CHANGELOG.md
# Changelog

## [0.7.0] - 2022-12-08
### Changed
- Upgraded to [django-ratelimit](https://github.com/jsocol/django-ratelimit) version 4.
  If you are using this, and your version is older, you will have to upgrade it.
### Added
- Client addresses are now obtained from a configurable source.
  See the documentation of the `BLACKLIST_ADDRESS_SOURCE` setting.
  Thanks to [@jimmybutton](https://github.com/jimmybutton).

### Upgrading
If you are not using `django-ratelimit`, you don't have to do anything.
Otherwise, follow [their instructions](https://django-ratelimit.readthedocs.io/en/stable/upgrading.html)
to upgrade to version 4.

## [0.6.1] - 2021-04-20
### Changed
- Correction to the previous changelog entry.

## [0.6.0] - 2021-04-20
### Changed
- Converted the middleware to a class.

### Upgrading
In your `MIDDLEWARE` in `settings.py`,
change `blacklist.middleware.blacklist_middleware` to `blacklist.middleware.BlacklistMiddleware`.

## [0.5.1] - 2021-03-05
### Changed
- Added a blank line for consistency. No other changes in this release.

## [0.5.0] - 2021-03-02
### Added
- Added logging toggle setting. Thanks to [@sweldon](https://github.com/sweldon).
### Fixed
- Allow custom user models. Thanks to [@richardARPANET](https://github.com/richardARPANET).

## [0.4.1] - 2020-05-10
### Changed
- Limited the length of comments in rules.
- Bulk deletion of rules from the database.

## [0.4.0] - 2020-05-10
### Added
- Reload rules periodically.
### Changed
- Major scalability improvements.

## [0.3.0] - 2019-12-14
### Added
- Setting to enable the automatic blacklisting of rate-limited clients (on by default).

## [0.2.2] - 2019-12-01
### Fixed
- Don't try to blacklist ratelimited clients if `BLACKLIST_ENABLE` is off.

## [0.2.1] - 2019-12-01
### Fixed
- Corrected the package version.
- Corrected the release date of version 0.2.0 in the changelog.

## [0.2.0] - 2019-12-01
### Added
- Render a custom error template, if configured.
- Setting to enable the blocking of blacklisted clients (on by default).

## [0.1.2] - 2019-06-22
### Fixed
- Install the `trim_blacklist` command with the package.
- Corrected the link to version 0.1.1 in the changelog.

## [0.1.1] - 2019-06-13
### Added
- Included the version of the package in the code.

### Fixed
- Corrected the link to version 0.1.0 in the changelog.

## [0.1.0] - 2019-06-13
### Added
- Initial version.

[0.7.0]: https://github.com/vsemionov/django-blacklist/compare/0.6.1...0.7.0
[0.6.1]: https://github.com/vsemionov/django-blacklist/compare/0.6.0...0.6.1
[0.6.0]: https://github.com/vsemionov/django-blacklist/compare/0.5.1...0.6.0
[0.5.1]: https://github.com/vsemionov/django-blacklist/compare/0.5.0...0.5.1
[0.5.0]: https://github.com/vsemionov/django-blacklist/compare/0.4.1...0.5.0
[0.4.1]: https://github.com/vsemionov/django-blacklist/compare/0.4.0...0.4.1
[0.4.0]: https://github.com/vsemionov/django-blacklist/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/vsemionov/django-blacklist/compare/0.2.2...0.3.0
[0.2.2]: https://github.com/vsemionov/django-blacklist/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/vsemionov/django-blacklist/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/vsemionov/django-blacklist/compare/0.1.2...0.2.0
[0.1.2]: https://github.com/vsemionov/django-blacklist/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/vsemionov/django-blacklist/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/vsemionov/django-blacklist/releases/tag/0.1.0


blacklist/models.py
import ipaddress
from datetime import timedelta

from django.db import models
from django.core import validators
from django.utils.timezone import now
from django.core.exceptions import ValidationError
from django.conf import settings


class Rule(models.Model):
    created = models.DateTimeField(auto_now_add=True)
    updated = models.DateTimeField(auto_now=True)

    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True, on_delete=models.CASCADE)

    address = models.GenericIPAddressField(null=True, blank=True)
    prefixlen = models.PositiveIntegerField(null=True, blank=True,
        validators=[validators.MinValueValidator(0), validators.MaxValueValidator(128)])

    duration = models.DurationField(validators=[validators.MinValueValidator(timedelta(0))])

    comments = models.TextField(max_length=2048, blank=True)

    def __str__(self):
        return str(self.id)

    def get_network(self):
        addr = self.address

        if addr:
            if self.prefixlen is not None:
                addr += f'/{self.prefixlen}'
            return ipaddress.ip_network(addr, strict=False)

        else:
            return None

    def get_expires(self):
        return self.created + self.duration

    def is_active(self):
        return now() < self.get_expires()

    def clean(self):
        if not self.address:
            # blank input produces empty string
            self.address = None

        if self.address is None and self.prefixlen is not None:
            raise ValidationError('Prefixlen without address.')

        try:
            self.get_network()

        except ValueError as ve:
            raise ValidationError(str(ve))

        if self.user is None and self.address is None:
            raise ValidationError('Neither user nor address provided.')

        if self.user is not None and self.address is not None:
            raise ValidationError('Both user and address provided.')


blacklist/ratelimit.py
import logging
import ipaddress
from functools import wraps

from django.conf import settings

from django_ratelimit.exceptions import Ratelimited

from .models import Rule
from .middleware import _get_client_address, _add_rule as _middleware_add_rule


logger = logging.getLogger(__name__)


def blacklist_ratelimited(duration, block=True):
    def decorator(fn):
        @wraps(fn)
        def wrapper(request, *args, **kwargs):
            if request.limited:
                if getattr(settings, 'BLACKLIST_ENABLE', True) \
                        and getattr(settings, 'BLACKLIST_RATELIMITED_ENABLE', True):
                    if user_duration and request.user.is_authenticated:
                        _create_user_rule(request, user_duration)

                    elif ip_duration:
                        _create_ip_rule(request, ip_duration)

                    else:
                        logger.warning('Unable to blacklist ratelimited client.')

                if block:
                    raise Ratelimited()

            return fn(request, *args, **kwargs)

        return wrapper

    if isinstance(duration, tuple):
        user_duration, ip_duration = duration
    else:
        user_duration = ip_duration = duration

    assert user_duration or ip_duration

    return decorator


def _create_user_rule(request, duration):
    user = request.user
    comments = _create_comments(request)
    rule = Rule(user=user, duration=duration, comments=comments)
    _add_rule(rule)


def _create_ip_rule(request, duration):
    addr = _get_client_address(request)
    ip = ipaddress.ip_address(addr)
    comments = _create_comments(request)
    rule = Rule(address=ip.compressed, duration=duration, comments=comments)
    _add_rule(rule)


def _create_comments(request):
    request_id = getattr(request, 'id', '')
    comments = 'Automatically blacklisted ratelimited client.\n' \
        f'Request ID: {request_id}\n' \
        f'Request line: {request.method} {request.get_full_path()}\n'
    return comments


def _add_rule(rule):
    rule.save()
    _middleware_add_rule(rule)


blacklist/migrations/0001_initial.py
# Generated by Django 2.2.12 on 2020-05-10 23:18

import datetime
from django.conf import settings
import django.core.validators
from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='Rule',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('address', models.GenericIPAddressField(blank=True, null=True)),
                ('prefixlen', models.PositiveIntegerField(blank=True, null=True, validators=[django.core.validators.MinValueValidator(0), django.core.validators.MaxValueValidator(128)])),
                ('duration', models.DurationField(validators=[django.core.validators.MinValueValidator(datetime.timedelta(0))])),
                ('comments', models.TextField(blank=True, max_length=2048)),
                ('user', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL)),
            ],
        ),
    ]


blacklist/management/commands/trim_blacklist.py
from datetime import timedelta

from django.core.management.base import BaseCommand
from django.db.models import F
from django.utils.timezone import now

from ...models import Rule


class Command(BaseCommand):
    help = 'Trims the blacklist'

    def add_arguments(self, parser):
        parser.add_argument('-c', '--created', type=int, default=0,
            help='created days ago; default: 0')
        parser.add_argument('-e', '--expired', type=int, default=0,
            help='expired days ago; default: 0')

    def handle(self, *args, **options):
        self.stdout.write('Deleting expired rules')

        created_age = timedelta(days=options['created'])
        expired_age = timedelta(days=options['expired'])

        current_time = now()

        rules = Rule.objects.filter(created__lte=(current_time - F('duration')))
        rules = rules.filter(created__lte=(current_time - created_age))
        rules = rules.filter(created__lte=(current_time - expired_age - F('duration')))

        deleted = rules.delete()
        num_deleted = deleted[1].get('%s.%s' % (Rule._meta.app_label, Rule._meta.object_name), 0)

        self.stdout.write(self.style.SUCCESS('Deleted %d rule(s).' % num_deleted))


blacklist/middleware.py
import ipaddress
import threading
from datetime import datetime, timedelta
from typing import Optional, Dict, Union
import logging

from django.core.exceptions import ImproperlyConfigured, SuspiciousOperation
from django.db.models import F, Max, DateTimeField
from django.utils.deprecation import MiddlewareMixin
from django.utils.module_loading import import_string
from django.utils.timezone import now
from django.shortcuts import render
from django.conf import settings

from .models import Rule


logger = logging.getLogger(__name__)


_RELOAD_PERIOD = timedelta(seconds=getattr(settings, 'BLACKLIST_RELOAD_PERIOD', 60))


_user_blacklist: Dict[int, datetime] = {}
_addr_blacklist: Dict[Optional[int], Dict[Union[ipaddress.IPv4Network, ipaddress.IPv6Network], datetime]] = {}

_loaded: Optional[datetime] = None

_blacklist_lock = threading.Lock()


class Blacklisted(SuspiciousOperation):
    pass


class BlacklistMiddleware(MiddlewareMixin):
    def process_request(self, request):
        if getattr(settings, 'BLACKLIST_ENABLE', True):
            current_time = now()

            if _needs_reload(current_time):
                _load_blacklist()

            try:
                _filter_client(request, current_time)

            except Blacklisted as exception:
                template_name = getattr(settings, 'BLACKLIST_TEMPLATE', None)

                if template_name:
                    if getattr(settings, 'BLACKLIST_LOGGING_ENABLE', True):
                        logger.warning(exception)

                    context = {'request': request, 'exception': exception}
                    return render(request, template_name, context, status=400)

                raise


def _filter_client(request, current_time):
    user = request.user
    user_id = user.id

    addr = _get_client_address(request)

    # no logging here, because the event will be logged either by the caller, or by django.request

    until = _user_blacklist.get(user_id)
    if until is not None and until > current_time:
        raise Blacklisted('Blacklisted user: %s' % user.username)

    for prefixlen, blacklist in _addr_blacklist.items():
        network = Rule(address=addr, prefixlen=prefixlen).get_network()

        until = blacklist.get(network)
        if until is not None and until > current_time:
            raise Blacklisted('Blacklisted address: %s' % addr)


def _get_client_address(request):
    source = getattr(settings, 'BLACKLIST_ADDRESS_SOURCE', 'REMOTE_ADDR')

    if source in request.META:
        addr = request.META[source]

    elif callable(source):
        addr = source(request)

    elif isinstance(source, str) and '.' in source:
        func = import_string(source)
        addr = func(request)

    else:
        raise ImproperlyConfigured(
            'Unable to obtain the client address. '
            'Please see the documentation of the BLACKLIST_ADDRESS_SOURCE setting.'
        )

    if not addr:
        raise ImproperlyConfigured(
            'The client address is empty. '
            'Please check the BLACKLIST_ADDRESS_SOURCE setting.'
        )

    return addr


def _needs_reload(current_time):
    return _loaded is None or current_time >= _loaded + _RELOAD_PERIOD


def _load_blacklist():
    global _loaded, _user_blacklist, _addr_blacklist

    with _blacklist_lock:
        current_time = now()

        if _needs_reload(current_time):
            until = Max(F('created') + F('duration'), output_field=DateTimeField())
            rules = Rule.objects.values('user_id', 'address', 'prefixlen').annotate(until=until)
            rules = rules.filter(until__gt=current_time)

            user_blacklist = {}
            addr_blacklist = {}

            for rule in rules:
                user_id = rule['user_id']
                prefixlen = rule['prefixlen']
                network = Rule(address=rule['address'], prefixlen=prefixlen).get_network()
                until = rule['until']

                _add_client(user_blacklist, addr_blacklist, user_id, prefixlen, network, until)

            _user_blacklist = user_blacklist
            _addr_blacklist = addr_blacklist
            _loaded = now()


def _add_client(user_blacklist, addr_blacklist, user_id, prefixlen, network, until):
    if user_id is not None:
        current_until = user_blacklist.get(user_id)
        if current_until is None or current_until < until:
            user_blacklist[user_id] = until

    if network is not None:
        blacklist = addr_blacklist.setdefault(prefixlen, {})
        current_until = blacklist.get(network)
        if current_until is None or current_until < until:
            blacklist[network] = until


def _add_rule(rule):
    global _addr_blacklist

    with _blacklist_lock:
        user_id = rule.user_id
        prefixlen = rule.prefixlen
        network = rule.get_network()
        until = rule.get_expires()

        addr_blacklist = _addr_blacklist.copy()
        _add_client(_user_blacklist, addr_blacklist, user_id, prefixlen, network, until)
        _addr_blacklist = addr_blacklist
